File: backend-python/app/schema_export.py
"""Export the canonical (Alembic-owned) schema as a normalized JSON snapshot.

This is the **Python side of the schema-parity check** (DA-8): it walks
``Base.metadata`` (populated by importing ``app.models``) and emits a
language-neutral ``{table: {column: {type, nullable}}}`` map. The TypeORM
entities in ``backend-ts`` emit the *same* normalized shape; a ``contracts/``
test deep-compares the two so the two backends can never drift on tables,
columns, or column types (money / datetime / enum / token included).

Column types are reduced to a small set of **canonical tokens** that are
meaningful across both Postgres-via-SQLAlchemy and Postgres-via-TypeORM:

* ``bigint``                 — 64-bit integer PK / FK
* ``text``                   — unbounded string (enums are text + CHECK)
* ``varchar(N)``             — bounded string
* ``money``                  — ``NUMERIC(14, 2)`` (decimal-string on the wire)
* ``percentage``             — bare ``NUMERIC`` (a number 0-100 on the wire)
* ``date``                   — calendar date
* ``timestamptz``            — timezone-aware datetime
* ``boolean``                — boolean
* ``bytea``                  — binary (encrypted Plaid token ciphertext)
* ``text[]``                 — text array (Plaid ``products``)

Run as a script to print the JSON snapshot to stdout (the parity harness shells
out to this): ``python -m app.schema_export``.
"""

from __future__ import annotations

import json
from typing import Any

from sqlalchemy import (
    ARRAY,
    BigInteger,
    Boolean,
    Date,
    DateTime,
    Integer,
    LargeBinary,
    Numeric,
    String,
    Text,
)
from sqlalchemy.sql.schema import Column

from app.db import Base

# Import the models so their tables register on ``Base.metadata``. The import is
# side-effecting; keep it after Base to avoid a circular import surprise.
import app.models  # noqa: F401,E402


def canonical_type(column: Column[Any]) -> str:
    """Reduce a SQLAlchemy column type to a cross-language canonical token."""
    coltype = column.type

    if isinstance(coltype, ARRAY):
        return f"{canonical_scalar(coltype.item_type)}[]"
    return canonical_scalar(coltype)


def canonical_scalar(coltype: Any) -> str:
    """Canonical token for a non-array (scalar) SQLAlchemy type."""
    # Order matters: BigInteger is an Integer subclass, so test it first.
    if isinstance(coltype, BigInteger):
        return "bigint"
    if isinstance(coltype, Integer):
        return "integer"
    if isinstance(coltype, Numeric):
        # NUMERIC(14, 2) is money; bare NUMERIC (no precision) is a percentage.
        if coltype.precision == 14 and coltype.scale == 2:
            return "money"
        if coltype.precision is None and coltype.scale is None:
            return "percentage"
        return f"numeric({coltype.precision},{coltype.scale})"
    if isinstance(coltype, DateTime):
        return "timestamptz" if coltype.timezone else "timestamp"
    if isinstance(coltype, Date):
        return "date"
    if isinstance(coltype, Boolean):
        return "boolean"
    if isinstance(coltype, LargeBinary):
        return "bytea"
    # Text is a String subclass with no length; a bounded String -> varchar(N).
    if isinstance(coltype, Text):
        return "text"
    if isinstance(coltype, String):
        return f"varchar({coltype.length})" if coltype.length is not None else "text"
    raise ValueError(f"Unmapped column type: {coltype!r}")


def export_schema() -> dict[str, dict[str, dict[str, Any]]]:
    """Return ``{table: {column: {"type", "nullable"}}}`` for every model table."""
    snapshot: dict[str, dict[str, dict[str, Any]]] = {}
    for table_name in sorted(Base.metadata.tables):
        table = Base.metadata.tables[table_name]
        columns: dict[str, dict[str, Any]] = {}
        for column in table.columns:
            columns[column.name] = {
                "type": canonical_type(column),
                "nullable": bool(column.nullable),
            }
        snapshot[table_name] = columns
    return snapshot


def main() -> None:
    """Print the normalized schema snapshot as JSON to stdout."""
    print(json.dumps(export_schema(), indent=2, sort_keys=True))


if __name__ == "__main__":
    main()

File: backend-ts/src/entities/schema-export.ts
/**
 * Export the TypeORM entity schema as a normalized JSON snapshot.
 *
 * This is the **TypeScript side of the schema-parity check** (DA-8). It builds
 * the TypeORM metadata for every entity WITHOUT connecting to a database
 * (`DataSource.buildMetadatas()` — no `initialize()`, so it runs in CI with no
 * Postgres) and emits the SAME language-neutral
 * `{ table: { column: { type, nullable } } }` map that
 * `backend-python/app/schema_export.py` produces from the Alembic-owned schema.
 *
 * A `contracts/` test deep-compares the two snapshots so the entities and the
 * Alembic head can never drift on tables, columns, or column types (money /
 * datetime / enum / token included).
 *
 * Canonical type tokens (must match schema_export.py):
 *   bigint · integer · text · varchar(N) · money · percentage · date ·
 *   timestamptz · timestamp · boolean · bytea · text[]
 */

import 'reflect-metadata';
import { DataSource } from 'typeorm';
import type { ColumnMetadata } from 'typeorm/metadata/ColumnMetadata';

import { ALL_ENTITIES } from './entities';

type ColumnSnapshot = { type: string; nullable: boolean };
type TableSnapshot = Record<string, ColumnSnapshot>;
export type SchemaSnapshot = Record<string, TableSnapshot>;

/** Reduce a TypeORM column's metadata to a cross-language canonical token. */
export function canonicalType(column: ColumnMetadata): string {
  const base = canonicalScalar(column);
  return column.isArray ? `${base}[]` : base;
}

function canonicalScalar(column: ColumnMetadata): string {
  // `column.type` is the normalized TypeORM/Postgres type (string or ctor).
  const raw = typeof column.type === 'string' ? column.type : column.type.name;
  const type = raw.toLowerCase();

  switch (type) {
    case 'bigint':
      return 'bigint';
    case 'integer':
    case 'int':
    case 'int4':
      return 'integer';
    case 'numeric':
    case 'decimal': {
      // numeric(14,2) is money; bare numeric (no precision) is a percentage.
      if (column.precision === 14 && column.scale === 2) return 'money';
      if (column.precision === undefined && column.scale === undefined) {
        return 'percentage';
      }
      return `numeric(${column.precision},${column.scale})`;
    }
    case 'timestamptz':
    case 'timestamp with time zone':
      return 'timestamptz';
    case 'timestamp':
    case 'timestamp without time zone':
      return 'timestamp';
    case 'date':
      return 'date';
    case 'boolean':
    case 'bool':
      return 'boolean';
    case 'bytea':
      return 'bytea';
    case 'text':
      return 'text';
    case 'varchar':
    case 'character varying':
      return column.length ? `varchar(${column.length})` : 'text';
    default:
      throw new Error(
        `Unmapped TypeORM column type "${raw}" on ${column.entityMetadata.tableName}.${column.databaseName}`,
      );
  }
}

/** Build entity metadata (no DB connection) and return the normalized snapshot. */
export async function exportSchema(): Promise<SchemaSnapshot> {
  // type: 'postgres' so types normalize to the Postgres dialect; no host/port
  // is contacted because we never call initialize().
  const dataSource = new DataSource({
    type: 'postgres',
    entities: ALL_ENTITIES,
    synchronize: false,
  });
  // Build metadata for all entities WITHOUT opening a connection. `initialize()`
  // would connect to Postgres; `buildMetadatas()` only constructs the in-memory
  // EntityMetadata graph (it is `protected` + async, hence the cast + await).
  // This keeps the exporter hermetic — it runs in CI with no database.
  await (
    dataSource as unknown as { buildMetadatas: () => Promise<void> }
  ).buildMetadatas();

  const snapshot: SchemaSnapshot = {};
  for (const meta of dataSource.entityMetadatas) {
    const columns: TableSnapshot = {};
    for (const column of meta.columns) {
      columns[column.databaseName] = {
        type: canonicalType(column),
        nullable: column.isNullable,
      };
    }
    // Sort keys so the JSON is stable/diffable.
    snapshot[meta.tableName] = Object.fromEntries(
      Object.entries(columns).sort(([a], [b]) => a.localeCompare(b)),
    );
  }
  // Sort tables too.
  return Object.fromEntries(
    Object.entries(snapshot).sort(([a], [b]) => a.localeCompare(b)),
  );
}

/** Print the normalized schema snapshot as JSON to stdout (the parity harness
 * shells out to this). Exported so it can be unit-tested. */
export async function main(): Promise<void> {
  const snapshot = await exportSchema();
  process.stdout.write(JSON.stringify(snapshot, null, 2) + '\n');
}

// Run as a script: `node dist/entities/schema-export.js` (the parity harness
// shells out to this). `require.main === module` is the CJS entry guard Nest
// builds to.
if (require.main === module) {
  void main();
}
